src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -265,75 +264,67 @@ public static bool FilterPackages(IEnumerable<FrameworkPackages> packages, ListP
+        /// <param name="listPackageArgs">List command args</param>
+        /// <returns>A dictionary where the key is the package id, and the value is a list of <see cref="IPackageSearchMetadata"/>.</returns>
+        private async Task<Dictionary<string, List<IPackageSearchMetadata>>> GetPackageMetadataAsync(
+            List<FrameworkPackages> targetFrameworks,
             ListPackageArgs listPackageArgs)
         {
-            //Unique Dictionary for packages and list of latest versions to handle different sources
-            var packagesVersionsDict = new Dictionary<string, IList<IPackageSearchMetadata>>();
-            AddPackagesToDict(packages, packagesVersionsDict);
+            List<string> allPackages = GetAllPackageIdentifiers(targetFrameworks, listPackageArgs.IncludeTransitive);
+            var packageMetadataById = new Dictionary<string, List<IPackageSearchMetadata>>(capacity: allPackages.Count);
 
-            //Prepare requests for each of the packages
-            var getLatestVersionsRequests = new List<Task>();
-            foreach (var package in packagesVersionsDict)
+            // We're (probably) going to make a bunch of HTTP requests, so limit to max 4 packages in parallel to avoid being unkind to servers.
+            // Note, each package will also run all sources in parallel, so the max concurrent HTTP requests is higher.
+            int taskCount = Math.Min(allPackages.Count, 4);
+            var tasks = new Task<KeyValuePair<string, List<IPackageSearchMetadata>>>[taskCount];
+
+            // ramp up throttling
+            int packageIndex;
+            for (packageIndex = 0; packageIndex < taskCount; packageIndex++)
             {
-                getLatestVersionsRequests.AddRange(
-                    PrepareLatestVersionsRequests(
-                        package.Key,
-                        listPackageArgs,
-                        packagesVersionsDict));
+                tasks[packageIndex] = GetPackageVersionsAsync(allPackages[packageIndex], listPackageArgs);
             }
 
-            // Make requests in parallel.
-            await RequestNuGetResourcesInParallelAsync(getLatestVersionsRequests);
+            // throttling steady state
+            while (packageIndex < allPackages.Count)
+            {
+                _ = await Task.WhenAny(tasks);
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    if (tasks[i].IsCompleted)
+                    {
+                        KeyValuePair<string, List<IPackageSearchMetadata>> result = await tasks[i];
+                        packageMetadataById[result.Key] = result.Value;
 
-            //Save latest versions within the InstalledPackageReference
-            await GetVersionsFromDictAsync(packages, packagesVersionsDict, listPackageArgs);
-        }
+                        tasks[i] = GetPackageVersionsAsync(allPackages[packageIndex++], listPackageArgs);
```

Comment:
IIRC we already have other throttling pattern code elsewhere. Should we reuse instead of introducing new one?

Reply:
yes, ideally. I spent a short time searching and didn't find anything useful.

There's a `SemaphoreThottle`, but I don't like the API, because it just has a `Wait()` (and I don't understand what the type provides over using `Semaphore` or `SemaphoreSlim` directly). But I don't want to start all tasks in parallel, and have the tasks internally throttle. I'd rather throttle the creation of the tasks that are being awaited. Uses less memory.

But I have a much higher tolerance for boilerplate code than most people seem to. To me, reusable code has two primary purposes. Firstly, to avoid implementation bugs. If someone else has already implemented it, hopefully it's bug free. Secondly, when requirements change, there's a single place to change and the entire product now implements the change.  This method is 30 lines, so hopefully fits on most peoples screens without scrolling, so not unreasonable to read and look for bugs. It doesn't contain business logic that needs to be applied in other areas of the product. So to me, it's not an important use case for reusing code 🤷